File: src/station/perturbations.py
"""Cross-domain perturbation harness (Phase-6 Step 8, P6.8) — cascades, no cascade code.

The Phase-3 :mod:`domains.biosphere.perturbations` discipline carried **cross-domain**:
a perturbation is a **scenario-layer intervention composed onto the already-assembled
station inputs**, never a core/domain change. Each shipped perturbation is a **cascade
with no cascade code** — a disturbance applied to **one** domain propagates into
**another** through a shared stock (or a shared *forcing*, #16) alone, while
conservation holds and ``rationed`` *behaves*. The harness composes *outside* the
``build_*`` builders, so ``git diff src/{simcore,domains}/`` stays empty and all twenty
existing goldens stay byte-identical; a perturbation run gets **no golden** (a
behavioural demonstration — the Phase-3 "diagnostics, no golden" precedent).

**The genuinely-new thing vs Phase-3.** Phase-3's three perturbations were
*single-domain* (all inside the biosphere). Here each perturbation crosses a domain
boundary:

* **brownout** (``solar_power`` cut) — Power SOC↓ ⇒ the Thermal node **cools** (less
  dissipation); deep/long enough it empties the battery ⇒ ``rationed > 0`` **emerges**
  (the *failure* cascade the exit criterion wants).
* **radiator failure** (:class:`ScaledFlow` throttling ``RadiatorReject``) — the node
  can no longer shed Power's real dissipation ⇒ it **heats**, T rises. Energy stays
  in-system (conserved); ``rationed == 0`` (a POOL accumulation, not a withdrawal
  shortfall).
* **atmosphere leak** (:class:`~domains.biosphere.perturbations.LeakFlow` on the shared
  cabin air) — the two gas pools **do not fail the same way** (below): a ``CARBON_POOL``
  leak drops Ci ⇒ the plant assimilates less + the scrubber does less; an ``O2_POOL``
  leak is absorbed by ``O2Makeup`` (``o2_supply`` effort↑, ``cabin_o2`` flat).
* **crew load spike** (``food_intake`` ×factor) — both ECLSS regulators work harder
  (``co2_removed↑`` + ``o2_supply↑``) and ``food_store`` depletes faster.
* **lighting failure** (``par`` **and** ``lamp_power`` → 0) — the biosphere's growth
  stalls **and** the battery is spared: the #16 lamp is **one** intervention with a
  photon leg (a forcing) and an energy leg (the ``Lamp`` flow's draw).

**The load-bearing finding (advisor, spike-CONFIRMED): the station regulators ERASE the
naive pool-level signature.** Unlike Phase-3's un-regulated chamber, every station gas
pool is regulated, so in every matter perturbation the *day-boundary* ``CARBON_POOL`` /
``O2_POOL`` / ``Ci`` return **identical to baseline** (the Step-3 regulator-erasure
physics, under disturbance). The emergent signature is regulator **effort**
(``co2_removed`` / ``o2_supply``) + the sinks (``LEAK_SINK``, biomass), **not** pool
level — and the two pools differ: ``CARBON_POOL`` is only *removed* (``CO2Scrubber`` is
first-order donor-controlled — it cannot push CO₂ *up*), so a leak genuinely lowers it
*within* the window; ``O2_POOL`` is *defended* (``O2Makeup`` is demand-controlled toward
a setpoint), so a leak surfaces as makeup effort. Tests assert on the effort/sink
signal, never the erased pool level.

**Three seam-types.** (1) **Forcing override** — reuse Phase-3's generic
:func:`~domains.biosphere.perturbations.window_override` /
:func:`~domains.biosphere.perturbations.with_forcing` (pure ``(n, dt)``
schedule/resolver transforms), plus :func:`window_scale` (the station's *multiplicative*
window, for scaling an existing schedule rather than forcing a constant). (2) **Added
leak flow** — reuse Phase-3's :class:`~domains.biosphere.perturbations.LeakFlow` +
:data:`~domains.biosphere.perturbations.LEAK_SINK` (a generic, composition-mirroring
``pool → sink``). (3) **Windowed flow-scaler** — the legitimately-new
:class:`ScaledFlow` (scale an existing flow's *whole* output by a windowed ``health ∈
[0, 1]`` so it stays internally balanced — the "arbitration scales the whole flow"
invariant, applied as a perturbation).

**Rejected — a ``Perturbation`` protocol.** Speculative generality for 5 perturbations
(the Phase-3 call); the small builder functions + two shared helpers here are leaner,
additive if a real multi-perturbation composition need appears. Pure stdlib +
``simcore`` + ``domains`` + ``station`` (the assembly layer imports domains freely; no
domain imports another).
"""

from dataclasses import dataclass, replace

from domains.biosphere.perturbations import (
    LEAK_SINK,
    LEAK_VAR,
    LeakFlow,
    window_override,
    with_forcing,
)
from domains.biosphere.stocks import PAR_VAR
from domains.crew.stocks import FOOD_INTAKE_VAR
from domains.power.stocks import SOLAR_POWER_VAR
from domains.thermal.system import RADIATOR_REJECT
from simcore.boundary import BOUNDARY_DOMAIN
from simcore.environment import Environment, Schedule, SourceResolver, constant
from simcore.flow import Flow, FlowResult, Leg
from simcore.ids import FlowId, StockId
from simcore.quantities import StockKind
from simcore.registry import Registry
from simcore.state import State, Stock
from station.flows import LAMP_POWER_VAR

# The radiator's windowed "health" forcing var (1.0 = nominal, 0.0 = total loss). Local
# to this module — never added to any domain's stock/forcing catalog, so baseline
# assembly never carries it and the existing goldens are untouched (the ``LEAK_VAR``
# discipline).
RADIATOR_HEALTH_VAR: str = "radiator_health"

# The station leak flow's id — station-owned (the leak reads a shared gas pool and
# writes a boundary sink), kept out of the biosphere registry so the sealed build's
# flow-id disjointness guard holds (:func:`station.sealed._assert_flow_ids_disjoint`).
STATION_LEAK: FlowId = FlowId("station.atmospheric_leak")


# --- the multiplicative window (the complement to window_override) --------------------


def window_scale(base: Schedule, *, start: int, end: int, factor: float) -> Schedule:
    """A ``Schedule`` returning ``factor · base`` on ``[start, end)``, else ``base``.

    The *multiplicative* sibling of Phase-3's :func:`window_override` (which forces a
    fixed value): a pure function of the integer step ``n`` (#14) that **scales** an
    existing, ``n``-varying schedule over the window — a brownout dims the diurnal
    ``solar_power`` half-sine by ``factor``, a crew load spike lifts ``food_intake`` by
    ``factor > 1``. ``factor = 1`` is a no-op (bit-identical); ``factor = 0``
    degenerates to a blackout (equivalent to ``window_override(base, value=0)``). ``dt``
    threads to ``base`` unchanged.
    """

    def schedule(n: int, dt: float) -> float:
        if start <= n < end:
            return factor * base(n, dt)
        return base(n, dt)

    return schedule


# --- forcing-override perturbations (no new flow / no structural change) --------------


def with_brownout(
    resolver: SourceResolver, *, start: int, end: int, factor: float = 0.0
) -> SourceResolver:
    """Scale the ``solar_power`` forcing by ``factor`` over ``[start, end)`` (Power).

    The cross-domain energy perturbation: over the window the diurnal solar supply is
    dimmed (``factor = 0`` is a full blackout), so the battery SOC falls (Power) **and**
    the heat Power dissipates into ``thermal.node`` drops, so the node **cools**
    (Thermal) — a cascade with no cascade code. A *short/shallow* window stays graceful
    (``rationed == 0``, SOC dips but > 0); a *deep/long* one empties the battery so
    ``LoadDraw`` cannot be met and ``rationed > 0`` **emerges** (the failure cascade,
    still conserving — the Euler backstop conserves as it rations). Runs on the diurnal
    single-rate ``run_station``.
    """
    base = resolver.forcings[SOLAR_POWER_VAR]
    return with_forcing(
        resolver,
        SOLAR_POWER_VAR,
        window_scale(base, start=start, end=end, factor=factor),
    )


def with_crew_load_spike(
    resolver: SourceResolver, *, start: int, end: int, factor: float = 2.0
) -> SourceResolver:
    """Scale crew ``food_intake`` by ``factor`` over ``[start, end)`` (Crew).

    The cross-domain crew perturbation: a raised metabolic food intake drives more
    respiration, so the cabin CO₂/O₂ loads jump — but the regulators absorb the pools,
    so the emergent signature is **regulator effort** (``CO2Scrubber`` removes more,
    ``O2Makeup`` supplies more) plus a **faster ``food_store`` drawdown**, not a
    pool-level shift (the day-boundary ``CARBON_POOL`` / ``O2_POOL`` return to
    setpoint). O₂ consumption is derived from food (RQ = 1, the ``CrewRespiration``
    merge), so scaling food scales the O₂ side too; the WATER echo (a proportional
    ``water_intake`` spike) is left off — food is the gas-loop lever. Runs on the short
    two-rate ``run_sealed`` (the fast registry's ``food_intake``).
    """
    base = resolver.forcings[FOOD_INTAKE_VAR]
    return with_forcing(
        resolver,
        FOOD_INTAKE_VAR,
        window_scale(base, start=start, end=end, factor=factor),
    )


def with_lighting_failure(
    bio_resolver: SourceResolver,
    fast_resolver: SourceResolver,
    *,
    start: int,
    end: int,
) -> tuple[SourceResolver, SourceResolver]:
    """Zero ``par`` **and** ``lamp_power`` over ``[start, end)`` (#16 lamp).

    The one **two-resolver** perturbation: the lamp is a single device whose failure has
    a *photon* leg (the biosphere's ``par`` **forcing**, in ``bio_resolver``) and an
    *energy* leg (the ``Lamp`` flow's ``lamp_power`` draw, in ``fast_resolver``) — a
    flow cannot tell forcing from a shared stock (#16), so both must be cut
    **together**. The cascade: PAR → 0 ⇒ FvCB's light term collapses ⇒ growth stalls
    (biomass below baseline); and the lamp draws no energy ⇒ the battery is **spared**
    (drains slower). Returns the ``(bio, fast)`` resolver pair. Runs on the short
    two-rate ``run_sealed``.
    """
    par = bio_resolver.forcings[PAR_VAR]
    lamp = fast_resolver.forcings[LAMP_POWER_VAR]
    new_bio = with_forcing(
        bio_resolver, PAR_VAR, window_override(par, start=start, end=end, value=0.0)
    )
    new_fast = with_forcing(
        fast_resolver,
        LAMP_POWER_VAR,
        window_override(lamp, start=start, end=end, value=0.0),
    )
    return new_bio, new_fast


# --- the windowed flow-scaler (the new third seam-type) -------------------------------


@dataclass(frozen=True)
class ScaledFlow:
    """Wrap a flow; multiply **all** its legs by a windowed ``health ∈ [0, 1]`` forcing.

    The legitimately-new third seam-type (radiator failure): a *degrade an existing
    process* perturbation. It scales the **whole flow** — every leg by the same
    ``env.get(health_var)`` — so the result stays internally balanced (``Σ (α·leg) = α·Σ
    leg = 0`` per quantity), the "arbitration scales the whole flow" invariant applied
    as a disturbance rather than a backstop. ``health = 1`` outside the window
    reproduces the wrapped flow **bit-identically** (``x · 1.0 == x`` in IEEE), so the
    perturbation is confined to the window exactly. The wrapped flow is unmodified
    (composition over inheritance); ``id`` / ``priority`` delegate, so the ``Registry``
    sorts it into the wrapped flow's slot (order-independence preserved).
    """

    inner: Flow
    health_var: str

    @property
    def id(self) -> FlowId:
        return self.inner.id

    @property
    def priority(self) -> int:
        return self.inner.priority

    def evaluate(self, snapshot: State, env: Environment, dt: float) -> FlowResult:
        alpha = env.get(self.health_var)
        result = self.inner.evaluate(snapshot, env, dt)
        return FlowResult(
            legs=tuple(Leg(leg.stock, leg.amount * alpha) for leg in result.legs)
        )


def with_radiator_failure(
    state: State,
    registry: Registry,
    resolver: SourceResolver,
    *,
    start: int,
    end: int,
    health: float = 0.0,
) -> tuple[Registry, SourceResolver]:
    """Throttle ``RadiatorReject`` to ``health`` over ``[start, end)`` (Thermal).

    Wraps the ``RADIATOR_REJECT`` flow in a :class:`ScaledFlow` reading a windowed
    :data:`RADIATOR_HEALTH_VAR` (``health`` inside the window, ``1.0`` outside) and
    rebuilds the registry over ``state.stocks`` (``Registry`` re-sorts by id, so append
    order is inert; aux processes are carried; ``state`` is read only for its stock
    mapping — the perturbation adds no stock). Over the window the radiator sheds only
    ``health ×`` its nominal rejection, so the node accumulates Power's real dissipation
    and **heats** — the emergent overheating cascade. Energy is conserved throughout
    (the scaled leg is still balanced — heat stays in the node instead of leaving to
    ``space``); ``rationed == 0`` (a POOL accumulation, not a withdrawal shortfall).
    Structural only (no new stock), so returns ``(registry, resolver)``. Runs on the
    single-rate ``run_station``.
    """
    flows: list[Flow] = [
        ScaledFlow(flow, RADIATOR_HEALTH_VAR) if flow.id == RADIATOR_REJECT else flow
        for flow in registry.flows
    ]
    new_registry = Registry(flows, state.stocks, aux_processes=registry.aux_processes)
    new_resolver = with_forcing(
        resolver,
        RADIATOR_HEALTH_VAR,
        window_override(constant(1.0), start=start, end=end, value=health),
    )
    return new_registry, new_resolver


# --- the added leak flow (two-registry sealed build) ----------------------------------


def with_station_leak(
    state: State,
    bio_registry: Registry,
    fast_registry: Registry,
    fast_resolver: SourceResolver,
    *,
    pool: StockId,
    k_leak: float,
    start: int,
    end: int,
) -> tuple[State, Registry, Registry, SourceResolver]:
    """Augment the sealed build with a windowed leak ``pool → LEAK_SINK`` (matter).

    The two-registry analogue of Phase-3's
    :func:`~domains.biosphere.perturbations.with_atmospheric_leak`. The sealed station
    has **two** registries over **one** shared stock dict, so:

    * a :data:`~domains.biosphere.perturbations.LEAK_SINK` BOUNDARY stock whose
      composition **mirrors ``pool``'s** is added to the shared stocks (a ``{C:1,O:2}``
      cabin-CO₂ pool vents CARBON **and** OXYGEN in balance);
    * **both** registries are rebuilt over the augmented stock dict (the biosphere-slow
      one preserving its ``thermal_time`` aux), and the :class:`LeakFlow` is appended to
      the **fast** registry (``dt = 60 s``, the same rate the cabin flows act on the
      shared pool; ``k_leak·dt < 1`` is trivial at 60 s). Its id (:data:`STATION_LEAK`)
      is kept **out of the biosphere registry**, so the sealed build's flow-id
      disjointness guard holds;
    * the :data:`~domains.biosphere.perturbations.LEAK_VAR` activation forcing is wired
      into the **fast** resolver as a windowed override (``1`` on ``[start, end)``,
      ``0`` else) — under ``substep`` the day count ``n`` is frozen within a day, so the
      window activates on whole master days (the sealed-station timing).

    The chamber interior's closure breaks over the window (mass leaves to ``LEAK_SINK``)
    but **total** mass (interior + sink) stays conserved — the leg is balanced, so a
    completed run (the driver's per-sub-step gate folds ``LEAK_SINK``) is itself the
    conservation proof. The emergent cascade is pool-specific: a ``CARBON_POOL`` leak
    lowers Ci within the window (biomass below baseline, scrubber does less), an
    ``O2_POOL`` leak is absorbed by ``O2Makeup`` (its ``o2_supply`` effort grows,
    ``cabin_o2`` stays flat).
    """
    pool_stock = state.stocks[pool]
    leak_sink = Stock(
        id=LEAK_SINK,
        domain=BOUNDARY_DOMAIN,
        quantity=pool_stock.quantity,
        unit=pool_stock.unit,
        amount=0.0,
        kind=StockKind.BOUNDARY,
        composition=dict(pool_stock.composition),
    )
    new_stocks = {**dict(state.stocks), LEAK_SINK: leak_sink}
    new_state = replace(state, stocks=new_stocks)
    leak = LeakFlow(
        STATION_LEAK,
        0,
        pool=pool,
        sink=LEAK_SINK,
        leak_var=LEAK_VAR,
        k_leak=k_leak,
    )
    new_bio = Registry(
        list(bio_registry.flows),
        new_stocks,
        aux_processes=bio_registry.aux_processes,
    )
    new_fast = Registry([*fast_registry.flows, leak], new_stocks)
    new_fast_resolver = with_forcing(
        fast_resolver,
        LEAK_VAR,
        window_override(constant(0.0), start=start, end=end, value=1.0),
    )
    return new_state, new_bio, new_fast, new_fast_resolver

File: rust/crates/station/tests/perturbations.rs
//! Phase-8 (P8.5) cross-domain perturbation cascades — the Rust port of the behavioural
//! asserts in `tests/test_station_perturbations.py`.
//!
//! Each perturbation is a **cascade with no cascade code**: a disturbance in one domain
//! propagates into another through a shared stock / shared forcing (#16), while conservation
//! holds (a completed run *is* the every-step proof — the gate lives in `step_report` /
//! the master-day driver) and `rationed` *behaves*. Direction-only asserts (never a
//! magnitude / `State == State` on a healthy run), each a perturbed-vs-baseline contrast.
//! No golden (the "diagnostics, no golden" precedent); the determinism re-runs are the
//! insurance.
//!
//! Two substrates: **energy** perturbations on the cheap single-rate diurnal station
//! (brownout, radiator failure); **matter** perturbations on a short two-rate sealed station
//! (leak, crew spike, lighting failure) with the window inside year 1 so the annual reset
//! never fires. The station regulators *erase* the pool level (P6.8), so matter cascades
//! assert regulator **effort** + sinks, not pool level.

use std::sync::OnceLock;

use domains::biosphere::perturbations::LEAK_SINK;
use domains::biosphere::stocks::{CARBON_POOL, LEAF_C, O2_POOL, ROOT_C, STEM_C, STORAGE_C};
use domains::crew::FOOD_STORE;
use domains::eclss::{CO2_REMOVED, O2_SUPPLY};
use domains::power::BATTERY;
use domains::thermal::{ThermalParams, NODE, SPACE};

use simcore::integrator::EulerIntegrator;
use simcore::state::State;

use station::driver::run_master_day;
use station::perturbations::{
    with_brownout, with_crew_load_spike, with_lighting_failure, with_radiator_failure,
    with_station_leak,
};
use station::scenario::{SealedStationScenario, HEAT_CLOSURE_SCENARIO};
use station::sealed::{build_sealed_station, sealed_bio_resolver, sealed_fast_resolver};
use station::system::{build_station, station_resolver};

// ============================ ENERGY: single-rate run_station ========================

const E_DAYS: u64 = 12;

fn e_spd() -> u64 {
    HEAT_CLOSURE_SCENARIO.power.steps_per_day
}
fn e_dt() -> f64 {
    HEAT_CLOSURE_SCENARIO.power.dt_seconds
}

fn node_temp(state: &State, thermal: &ThermalParams) -> f64 {
    thermal.space_temperature + state.stocks[NODE].amount / thermal.heat_capacity
}

/// Build the baseline Power → Thermal diurnal trajectory + the loaded params.
fn energy_baseline() -> (Vec<State>, ThermalParams) {
    let charge = domains::params::charge();
    let thermal = domains::params::thermal();
    let (state, reg) = build_station(&charge, &thermal, &HEAT_CLOSURE_SCENARIO, None).unwrap();
    let resolver = station_resolver(&charge, &HEAT_CLOSURE_SCENARIO).unwrap();
    let integ = EulerIntegrator::new(reg);
    let mut states = Vec::new();
    let (_f, rationed, events) = station::run_station(
        &integ,
        state,
        &resolver,
        e_dt(),
        E_DAYS * e_spd(),
        &mut |s| states.push(s.clone()),
    )
    .unwrap();
    assert_eq!(rationed, 0);
    assert!(events.is_empty());
    (states, thermal)
}

#[test]
fn brownout_graceful_cools_node_without_rationing() {
    let (base_states, thermal) = energy_baseline();
    let charge = domains::params::charge();
    let (state, reg) = build_station(&charge, &thermal, &HEAT_CLOSURE_SCENARIO, None).unwrap();
    let resolver = station_resolver(&charge, &HEAT_CLOSURE_SCENARIO).unwrap();
    // A short/shallow afternoon dimming: SOC dips below baseline and the node cools, no
    // rationing (the Power→Thermal cascade, graceful arm).
    let perturbed = with_brownout(resolver, 85, 90, 0.5).unwrap();
    let integ = EulerIntegrator::new(reg);
    let mut states = Vec::new();
    let (_f, rationed, events) = station::run_station(
        &integ,
        state,
        &perturbed,
        e_dt(),
        E_DAYS * e_spd(),
        &mut |s| states.push(s.clone()),
    )
    .unwrap();
    assert_eq!(rationed, 0);
    assert!(events.is_empty());

    let base_soc_min = base_states.iter().map(|s| s.stocks[BATTERY].amount).fold(f64::INFINITY, f64::min);
    let soc_min = states.iter().map(|s| s.stocks[BATTERY].amount).fold(f64::INFINITY, f64::min);
    assert!(soc_min < base_soc_min && soc_min > 0.0, "SOC dips but never empties");

    let base_t_min = base_states.iter().map(|s| node_temp(s, &thermal)).fold(f64::INFINITY, f64::min);
    let t_min = states.iter().map(|s| node_temp(s, &thermal)).fold(f64::INFINITY, f64::min);
    assert!(t_min < base_t_min, "node cools — the cross-domain cascade");
}

#[test]
fn brownout_deep_emerges_rationing_still_conserving() {
    // The FAILURE cascade: a multi-day full blackout empties the battery so LoadDraw cannot
    // be met → rationed > 0 EMERGES (bounded), yet ENERGY still conserves (a completed run
    // is the every-step proof — the Euler backstop conserves as it rations).
    let charge = domains::params::charge();
    let thermal = domains::params::thermal();
    let (state, reg) = build_station(&charge, &thermal, &HEAT_CLOSURE_SCENARIO, None).unwrap();
    let resolver = station_resolver(&charge, &HEAT_CLOSURE_SCENARIO).unwrap();
    let perturbed = with_brownout(resolver, 2 * e_spd(), 8 * e_spd(), 0.0).unwrap();
    let integ = EulerIntegrator::new(reg);
    let (_f, rationed, events) = station::run_station(
        &integ,
        state,
        &perturbed,
        e_dt(),
        E_DAYS * e_spd(),
        &mut |_| {},
    )
    .unwrap();
    assert!(rationed > 0, "the emergent failure signature");
    assert!(rationed < E_DAYS * e_spd(), "bounded (not every step rations)");
    assert!(events.is_empty());
}

#[test]
fn radiator_failure_heats_node_conserving_no_rationing() {
    let (base_states, thermal) = energy_baseline();
    let charge = domains::params::charge();
    let (state, reg) = build_station(&charge, &thermal, &HEAT_CLOSURE_SCENARIO, None).unwrap();
    let resolver = station_resolver(&charge, &HEAT_CLOSURE_SCENARIO).unwrap();
    // Throttle RadiatorReject to 0 over a window → Power's real dissipation piles up in the
    // node → it HEATS (peak T above baseline). Energy stays in-system (conserved); the sink
    // stays monotonic; rationed == 0 (a POOL accumulation, not a withdrawal shortfall).
    let (preg, pres) =
        with_radiator_failure(&state, reg, resolver, 3 * e_spd(), 9 * e_spd(), 0.0).unwrap();
    let integ = EulerIntegrator::new(preg);
    let mut states = Vec::new();
    let (_f, rationed, events) = station::run_station(
        &integ,
        state,
        &pres,
        e_dt(),
        E_DAYS * e_spd(),
        &mut |s| states.push(s.clone()),
    )
    .unwrap();
    assert_eq!(rationed, 0);
    assert!(events.is_empty());

    let base_t_max = base_states.iter().map(|s| node_temp(s, &thermal)).fold(f64::NEG_INFINITY, f64::max);
    let t_max = states.iter().map(|s| node_temp(s, &thermal)).fold(f64::NEG_INFINITY, f64::max);
    assert!(t_max > base_t_max, "node overheats — the cross-domain cascade");
    // The space sink is monotonic (heat never un-radiates).
    for w in states.windows(2) {
        assert!(w[1].stocks[SPACE].amount >= w[0].stocks[SPACE].amount - 1e-6);
    }
}

#[test]
fn inspection_is_truthful_under_deep_brownout_rationing() {
    // The P8.4 seam discharged on a REAL palette scenario (advisor): a deep brownout empties
    // the battery so LoadDraw rations (`scale < 1`) on the single-rate `station` entry — the
    // reachable rationing case. At every step, inspect the pre-step state and confirm
    // `before + applied_delta == after` for every stock (the scale-aware identity holds under
    // rationing, unlike the raw-leg sum), and that rationing actually occurred.
    let charge = domains::params::charge();
    let thermal = domains::params::thermal();
    let (mut state, reg) = build_station(&charge, &thermal, &HEAT_CLOSURE_SCENARIO, None).unwrap();
    let resolver = station_resolver(&charge, &HEAT_CLOSURE_SCENARIO).unwrap();
    let perturbed = with_brownout(resolver, 2 * e_spd(), 8 * e_spd(), 0.0).unwrap();
    let integ = EulerIntegrator::new(reg);

    let mut saw_rationing = false;
    for _ in 0..(E_DAYS * e_spd()) {
        let insp =
            station::inspection::inspect_flows(integ.registry(), &state, &perturbed, e_dt())
                .unwrap();
        let report = integ.step_report(&state, &perturbed, e_dt()).unwrap();
        if report.rationed > 0 {
            saw_rationing = true;
            // A rationed step must show at least one throttled flow in the inspection.
            assert!(insp.flows.iter().any(|f| f.scale < 1.0));
        }
        for (sid, before) in &state.stocks {
            let after = report.state.stocks[sid].amount;
            let applied = insp.applied_delta(sid);
            assert!(
                (before.amount + applied - after).abs() <= 1e-9 * after.abs() + 1e-9,
                "inspection lied about {sid} under rationing: {} + {applied} != {after}",
                before.amount
            );
        }
        state = report.state;
    }
    assert!(saw_rationing, "deep brownout should drive LoadDraw into rationing");
}

#[test]
fn radiator_failure_outside_window_is_baseline() {
    // The ScaledFlow bit-identity: health == 1 outside the window reproduces the wrapped flow
    // EXACTLY (x·1.0 == x), so a failure window the run never reaches gives the byte-identical
    // baseline trajectory.
    let (base_states, thermal) = energy_baseline();
    let charge = domains::params::charge();
    let (state, reg) = build_station(&charge, &thermal, &HEAT_CLOSURE_SCENARIO, None).unwrap();
    let resolver = station_resolver(&charge, &HEAT_CLOSURE_SCENARIO).unwrap();
    let (preg, pres) =
        with_radiator_failure(&state, reg, resolver, 100 * e_spd(), 101 * e_spd(), 0.0).unwrap();
    let integ = EulerIntegrator::new(preg);
    let (final_state, _r, _e) = station::run_station(
        &integ,
        state,
        &pres,
        e_dt(),
        E_DAYS * e_spd(),
        &mut |_| {},
    )
    .unwrap();
    assert_eq!(&final_state, base_states.last().unwrap());
}

// ============================ MATTER: short two-rate run_master_day ===================

const M_DAYS: usize = 8;
const M_START: u64 = 2;
const M_END: u64 = 7; // window (master days) — inside year 1
const K_LEAK: f64 = 1.0e-3;

// season_days >> the run horizon ⇒ the annual reset never fires (window stays in year 1).
fn matter_scenario() -> SealedStationScenario {
    SealedStationScenario {
        years: 1,
        season_days: 305,
        ..station::scenario::sealed_station_scenario()
    }
}

fn biomass(state: &State) -> f64 {
    [LEAF_C, STEM_C, ROOT_C, STORAGE_C]
        .iter()
        .map(|s| state.stocks[*s].amount)
        .sum()
}

/// Build the sealed pieces (`false, false` — the palette `sealed` config, no harvest / open
/// feces), matching the session-parity conventions.
fn sealed_build() -> (State, simcore::registry::Registry, simcore::registry::Registry) {
    let scn = matter_scenario();
    build_sealed_station(
        &domains::params::charge(),
        &domains::params::thermal(),
        &domains::params::crew(),
        &domains::params::eclss(),
        &station::params::water_recovery(),
        &station::params::lamp(),
        &station::params::harvest(),
        &scn,
        false,
        false,
    )
    .unwrap()
}

fn sealed_run(
    state: State,
    bio_reg: simcore::registry::Registry,
    fast_reg: simcore::registry::Registry,
    bio_res: &simcore::environment::SourceResolver,
    fast_res: &simcore::environment::SourceResolver,
) -> (Vec<State>, u64, Vec<simcore::events::Event>) {
    let scn = matter_scenario();
    run_master_day(
        &EulerIntegrator::new(bio_reg),
        &EulerIntegrator::new(fast_reg),
        state,
        bio_res,
        fast_res,
        M_DAYS,
        scn.steps_per_day,
        scn.bio_dt,
        scn.cabin_dt,
        None,
    )
    .unwrap()
}

/// The unperturbed short sealed final state — cached so each matter cascade pays one run.
fn sealed_baseline() -> &'static State {
    static BASE: OnceLock<State> = OnceLock::new();
    BASE.get_or_init(|| {
        let scn = matter_scenario();
        let (state, bio_reg, fast_reg) = sealed_build();
        let bio_res = sealed_bio_resolver(&station::params::lamp(), &scn).unwrap();
        let fast_res = sealed_fast_resolver(&domains::params::charge(), &scn).unwrap();
        let (states, rationed, events) =
            sealed_run(state, bio_reg, fast_reg, &bio_res, &fast_res);
        assert_eq!(rationed, 0);
        assert!(events.is_empty());
        states.last().unwrap().clone()
    })
}

fn run_carbon_leak(pool: &str) -> State {
    let scn = matter_scenario();
    let (state, bio_reg, fast_reg) = sealed_build();
    let fast_res = sealed_fast_resolver(&domains::params::charge(), &scn).unwrap();
    let bio_res = sealed_bio_resolver(&station::params::lamp(), &scn).unwrap();
    let (state, bio_reg, fast_reg, fast_res) = with_station_leak(
        &state, bio_reg, fast_reg, fast_res, pool, K_LEAK, M_START, M_END,
    )
    .unwrap();
    let (states, rationed, events) = sealed_run(state, bio_reg, fast_reg, &bio_res, &fast_res);
    assert_eq!(rationed, 0);
    assert!(events.is_empty());
    states.last().unwrap().clone()
}

#[test]
fn carbon_leak_lowers_biomass_and_scrubber_effort() {
    // CARBON_POOL is only-removed (the scrubber cannot push it up): the leak lowers Ci within
    // the window, so the plant assimilates LESS (biomass below baseline) AND the scrubber does
    // LESS work (leaked CO₂ was not there to remove). The leak-sink strictly accumulates.
    let baseline = sealed_baseline();
    let leaked = run_carbon_leak(CARBON_POOL);
    assert!(biomass(&leaked) < biomass(baseline));
    assert!(leaked.stocks[CO2_REMOVED].amount < baseline.stocks[CO2_REMOVED].amount);
    assert!(leaked.stocks[LEAK_SINK].amount > 0.0);
}

#[test]
fn o2_leak_is_absorbed_by_makeup_effort() {
    // O2_POOL is DEFENDED (O2Makeup is demand-controlled), so — unlike CARBON — the leak
    // surfaces as makeup EFFORT, not a pool/biology change: o2_supply supplies strictly MORE
    // (its cumulative bookkeeping runs further negative), the plant is UNTOUCHED, and the
    // leak-sink accumulates. The two pools fail differently.
    let baseline = sealed_baseline();
    let leaked = run_carbon_leak(O2_POOL);
    assert!(leaked.stocks[O2_SUPPLY].amount < baseline.stocks[O2_SUPPLY].amount);
    let rel = (biomass(&leaked) - biomass(baseline)).abs() / biomass(baseline).abs();
    assert!(rel < 1e-6, "biomass ≈ baseline (rel {rel})");
    assert!(leaked.stocks[LEAK_SINK].amount > 0.0);
}

#[test]
fn crew_spike_raises_regulator_effort_and_drains_food() {
    // A doubled food intake drives more respiration → BOTH regulators work harder (co2_removed
    // up AND o2_supply further negative) and food_store depletes faster. The gas pools are held
    // at setpoint by the regulators, so the signature is EFFORT, not level.
    let baseline = sealed_baseline();
    let scn = matter_scenario();
    let (state, bio_reg, fast_reg) = sealed_build();
    let bio_res = sealed_bio_resolver(&station::params::lamp(), &scn).unwrap();
    let fast_res = sealed_fast_resolver(&domains::params::charge(), &scn).unwrap();
    let fast_res = with_crew_load_spike(fast_res, M_START, M_END, 2.0).unwrap();
    let (states, rationed, events) = sealed_run(state, bio_reg, fast_reg, &bio_res, &fast_res);
    assert_eq!(rationed, 0);
    assert!(events.is_empty());
    let spiked = states.last().unwrap();

    assert!(spiked.stocks[CO2_REMOVED].amount > baseline.stocks[CO2_REMOVED].amount);
    assert!(spiked.stocks[O2_SUPPLY].amount < baseline.stocks[O2_SUPPLY].amount);
    assert!(spiked.stocks[FOOD_STORE].amount < baseline.stocks[FOOD_STORE].amount);
    // Regulator-erasure: the day-boundary pools return to the SAME setpoint as baseline.
    for pool in [CARBON_POOL, O2_POOL] {
        let rel = (spiked.stocks[pool].amount - baseline.stocks[pool].amount).abs()
            / baseline.stocks[pool].amount.abs();
        assert!(rel < 1e-6, "{pool} returns to setpoint (rel {rel})");
    }
}

#[test]
fn lighting_failure_stalls_growth_and_spares_battery() {
    // The #16 lamp is ONE device with two legs: cutting PAR (a forcing) stalls growth (biomass
    // below baseline), and cutting the lamp draw (the Lamp flow's energy) SPARES the battery
    // (it drains slower ⇒ higher SOC). One intervention, a cascade in both directions.
    let baseline = sealed_baseline();
    let scn = matter_scenario();
    let (state, bio_reg, fast_reg) = sealed_build();
    let bio_res = sealed_bio_resolver(&station::params::lamp(), &scn).unwrap();
    let fast_res = sealed_fast_resolver(&domains::params::charge(), &scn).unwrap();
    let (bio_res, fast_res) = with_lighting_failure(bio_res, fast_res, M_START, M_END).unwrap();
    let (states, rationed, events) = sealed_run(state, bio_reg, fast_reg, &bio_res, &fast_res);
    assert_eq!(rationed, 0);
    assert!(events.is_empty());
    let failed = states.last().unwrap();

    assert!(biomass(failed) < biomass(baseline), "growth stalls");
    assert!(
        failed.stocks[BATTERY].amount > baseline.stocks[BATTERY].amount,
        "battery spared"
    );
    // Regulator-erasure: the scrubber holds CARBON_POOL at setpoint despite less assimilation.
    let rel = (failed.stocks[CARBON_POOL].amount - baseline.stocks[CARBON_POOL].amount).abs()
        / baseline.stocks[CARBON_POOL].amount.abs();
    assert!(rel < 1e-6, "carbon pool returns to setpoint (rel {rel})");
}

#[test]
fn matter_perturbation_is_deterministic() {
    // The no-golden insurance (matter side): a perturbed two-rate sealed run is bit-identical
    // on re-run (stands in for the absent golden).
    let a = run_carbon_leak(CARBON_POOL);
    let b = run_carbon_leak(CARBON_POOL);
    assert_eq!(a, b);
}
